packages/cli/src/cms/api/middlewares/sunset.middleware.ts

```diff
@@ -0,0 +1,68 @@
+import type { Response } from 'node-fetch';
+
+export type SunsetWarning = { sunsetDate: Date; isSunsetExpired: boolean; warning: string };
+export type SunsetWarningsBuffer = SunsetWarning[];
+
+type SunsetMiddlewareOptions =
+  | {
+      bufferSunsets: true;
+    }
+  | {
+      onSunset: (arg: SunsetWarning) => void;
+    };
+
+export function createApiSunsetMiddleware(options: SunsetMiddlewareOptions) {
+  const buffer: SunsetWarningsBuffer = [];
+
+  function getSunsetDate(response: Response): number | undefined {
+    let sunsetDate = response.headers.get('sunset');
+
+    if (sunsetDate) {
+      return Date.parse(sunsetDate);
+    }
+
+    sunsetDate = response.headers.get('Sunset');
+
+    if (sunsetDate) {
+      return Date.parse(sunsetDate);
+    }
+
+    return;
+  }
+
+  return {
+    middleware: (response: Response) => {
+      const sunsetTime = getSunsetDate(response);
+
+      if (sunsetTime) {
+        const sunsetDate = new Date(sunsetTime);
+        let sunsetWarning: SunsetWarning;
+
+        if (sunsetTime > Date.now()) {
+          sunsetWarning = {
+            sunsetDate,
+            isSunsetExpired: false,
+            warning: `Endpoint ${
+              response.url
+            } is deprecated for removal on ${sunsetDate.toISOString()}`,
+          };
+        } else {
+          sunsetWarning = {
+            sunsetDate,
+            isSunsetExpired: true,
+            warning: `Endpoint ${
+              response.url
+            } was deprecated for removal on ${sunsetDate.toISOString()} and could be removed AT ANY TIME`,
```

Comment:
Please, never use upper case in user-facing content unless there is danger to life (or data).

```suggestion
            } support expired on ${sunsetDate.toISOString()} and will be removed.`,
```